internal/broker/broker.go

```diff
@@ -638,6 +631,37 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
 	return AuthGranted, userInfoMessage{UserInfo: authInfo.UserInfo}
 }
 
+// userNameIsAllowed checks whether the user's username is allowed to access the machine.
+func (b *Broker) userNameIsAllowed(userName string) bool {
+	// The user is allowed to login if:
+	// - ALL users are allowed
+	// - the user's name is in the list of allowed_users
+	// - the user is the owner of the machine
+	allowed := false
+	if b.cfg.userConfig.AllUsersAllowed() {
+		allowed = true
+	} else if b.cfg.userConfig.IsUserAllowed(userName) {
+		allowed = true
+	} else if b.cfg.userConfig.OwnerUserAllowed() {
+		// If owner is undefined, then the first user to login is considered the owner
+		if b.cfg.userConfig.OwnerIsUnset() || *b.cfg.userConfig.Owner() == userName {
+			allowed = true
+		}
+	}
```

Comment:
I feel this code can be put in a single function where you follow the happy-path better with early returns.